service/matching/matcher_test.go

```diff
@@ -210,55 +210,73 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		_, err := t.rootMatcher.Poll(ctx, &pollMetadata{})
+		t.Assert().NoError(err)
+		cancel()
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+
+	go func() {
+		historyTask.responseC <- nil
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+	happened, err := t.rootMatcher.Offer(ctx, historyTask)
+	t.True(happened)
+	t.Nil(err)
+}
+
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
 	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 
 	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	t.validateSyncMatchWhenNoBacklog(ctx, historyTask)
 
 	intruptC := make(chan struct{})
-	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
 
-	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
-	t.Nil(err)
-	t.False(happened) // sync match did not happen, but we called the forwarder client
+	// task waits for a local poller
+	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
+	go t.rootMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
+	time.Sleep(2 * time.Millisecond)
 
 	// should not allow sync match when there is an old task in backlog
-	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
-	go t.childMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
-	t.False(happened)
-	t.Nil(err)
 
-	// poll both tasks
-	task, _ := t.childMatcher.Poll(ctx, &pollMetadata{})
-	t.NotNil(task)
-	task, _ = t.childMatcher.Poll(ctx, &pollMetadata{})
+	backlogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
+	backlogTask.forwardInfo = &taskqueuespb.TaskForwardInfo{
```

Comment:
Adding `forwardInfo` to replicate a task being forwarded from the child partition. This field is required to be non-nil for the matcher to offer this task locally to a poller, which is what we want.

Reply:
let's add this comment to the test itself for future readers.